### PR TITLE
Refactor dial worker loop into an object and fix bug

### LINKED ISSUE #314: growing goroutine
More and more goroutines, more and more memory usage。

### PR DESCRIPTION
**Pull request by vyzo**
The dial worker loop has been refactored into an object to aid testability.
And more importantly, a goroutine leak is fixed by dropping the unnecessary active dial tracking logic.

Closes #314

### CODE REVIEW

File: dial_worker.go [hunk 1]
```diff
@@ -78,8 +79,8 @@ func (w *dialWorker) loop() {
 	defer w.wg.Done()
 	defer w.s.limiter.clearAllPeerDials(w.peer)
 
-	triggerNow := make(chan struct{})
-	close(triggerNow)
+	ready := make(chan struct{})
```
Comment (marten-seemann) [inline dial_worker.go:82]: What does `ready` do? Why is is closed immediately? Can we have a comment plz?

File: dial_worker.go [hunk 2]
```diff
@@ -57,9 +57,10 @@ type dialWorker struct {
 	nextDial    []ma.Multiaddr
 	triggerDial <-chan struct{}
 
+	complete <-chan struct{}
```
Comment (marten-seemann) [inline dial_worker.go:60]: Can we have a comment what this chan does?

File: dial_worker.go (hunk 1 above)
Comment (vyzo) [reply dial_worker.go:82]: it's a channel signalling immediate action in the loop; but yes, can haz comment.

File: dial_worker.go (hunk 2 above)
Comment (vyzo) [reply dial_worker.go:60]: sure!

Comment (vyzo) [reply dial_worker.go:63]: done.

File: dial_worker.go (hunk 1 above)
Comment (vyzo) [reply dial_worker.go:86]: added comment.

**Comment by vyzo** [comment]
squashed some commits to clean up history.

**Review by marten-seemann — APPROVED** [review]

**Pull request merged**